Refactor: testable controllers, simpler cloudformation abstraction

chrisfarms opened the pull request:
## What

Refactor the cloudformation controller abstraction

## Why

* So that we can easily write integration tests 
* So that we can mock out only the bare minimum AWS client functions
* So that we reduce the boilerplate required for creating cloudformation backed resource types

## :warning: breaking changes

* **Postgres password now managed by secretsmanager** which means existing postgres stacks will not update correctly and will need to be destroyed.
* **Finalizer name change**: since the majority of the cloudformation provisioning code is now shared between all instances of cloudformation controllers, all these resources now share a common finalizer name. This means that after deploying, any existing resources should be _manually kubectl edited to remove their current finalizer string_ which will trigger the controller to put back the new one.
* **Shape of the resource status changed**: all resources now share a common Status field to reduce boilerplate. This should not have any impact unless someone is processing the Status fields for some reason

Commit:
avoid fetching test deps if layer cached

to speed up build/test loop locally we can ensure that all the deps are
fetched ahead of the COPY ... but generate requires some things that go
mod download doesn't fetch ... this runs make generate to fetch them
so repeated build runs are faster

Commit:
implements cloudformation controller interface

Previously implementing cloudformation-based services required:

* creating a kubernetes object spec-type (+ crd)
* creating a cloudformation template-type unique to the spec-type
* creating a controller-type unique to both the object-spec type and the
  template-type to tie them together and knows how to extract values in
  to secrets and deal with the reconciliation loops, making use of a
  shared reconciler-func where possible.

Problems with this structure:

* This required quite a bit of boilerplate code that was very similar
  between implementations of the tepmlate-types and the controller-types
* Very slight differences in the boilerplate controller code made it
  easy to miss differing behaviour
* The interface for the shared reconciler code was leaking it's
  implementation requiring the controllers to perform fairly complex
  handling of the responses.
* The controllers were difficult to test without mocking out the entire
  reconciler (which itself contained logic that should be tested)

To address these issues, with an aim of making the controller code more
testable, the following interfaces where created...

cloudformation.Stack:

Any type that implements cloudformation.Stack can produce a
cloudformation template of itself, ready to be consumed by the
cloudformation client/controller.

cloudformation.StackPolicyAssigner:

A StackPolicyAssigner is an Stack with an additional method for mapping
a controller provided role-name into it's template (required because of
the current implementation of Principals/Roles.

cloudformation.StackSecretOutputter

A StackSecretOutputter is a Stack with an additional method that can
name a kubernets Secret that will be created/updated with any
cloudformation template outputs.

With these interfaces it was possible to write a generic controller
reconciler for any type that implements those interfaces, and means that
we no longer need to have seperate boilerplate types for
specs/controllers/templates. We can now direcly implement Stack
interfaces on spec types (saving the mapping of fields) and can
instaniate cloudformation.Controller for a given type (saving the need
to write boilerplate controllers).

A more generic interface object.Service and object.Status (intented to
be implmented by more object types than just cloudformation-based ones)
allow the sharing of the Status fields on objects, further reducing the
boilerplate required when implementing a new service type.

With the controllers fully generic, it was possible to create an
interface version of the AWS client (sdk.Client) which exposes only the
AWS SDK methods we make use of. Enabling mocking the client for use in
tests.

Integration tests have been added that can be run either in a mocked
mode (AWS_INTEGRATION=false) or against a real AWS environment
(AWS_INTEGRATION=true). The tests spin up a real kubernetes control
plane, spin up a controller manager, add the controllers to the
controller manager (in a way close to how the manager really starts up)
and then attempts to perform a create/update/delete of the resources
under test.

Examples of running the integration test can be found in
hack/test_integration.sh

Krenair commented on components/service-operator/apis/database/v1beta1/postgres_types.go:
```diff
@@ -16,60 +16,280 @@ limitations under the License.
+
+	template.Outputs[PostgresPasswordOutputName] = map[string]interface{}{
+		"Description": "Postgres master password",
+		"Value":       cloudformation.Ref(PostgresPassword),
```
We can't output this from a CloudFormation stack. The postgres handling in the service operator should only get this (and possibly the username) from the existing k8s Secret, to update into the new secret when we have outputs like endpoint and port.

chrisfarms replied:
Excellent point, thanks @Krenair  ...

I' think I can rejig the logic so that `Secrets` are not populated from cloudformation Outputs but by requesting a list of Secret Manager keys.

This would have the following benefits:

* Can still be implemented as a generic interface
* Will mean that secret generation can be removed from the operator entirely
* Will mean we could take advantage of rotation features of cloudformation/secret-manager in future

... I'll have a go at this now.

Krenair replied:
Well we still need to generate the values when we create the stack, but I'm sure this can be done with a generic interface too.

chrisfarms replied:
I addressed this by using secretsmanager :+1:

Krenair reviewed (dismissed):
security consideration that needs to be addressed before merge

chrisfarms commented:
Note to self: manager will now need some more IAM permissions (secretsmanager GetSecretValue)

Commit:
note about running tests

issyl0 commented:
Could you please update the commit messages, particularly 22cc498, with more context (what's on hte PR description at least)? A commit with stats like `+5,716 −2,294` and no way to `git blame` and see an explanation isn't useful for people in the future.

blairboy362 reviewed (commented):
Some minor comments.

chrisfarms commented on components/service-operator/internal/aws/cloudformation/controller.go:
```diff
@@ -0,0 +1,332 @@
+func (r *Controller) getRoleParams(ctx context.Context, o Stack) ([]*Parameter, error) {
+	serviceThatRequiresIAM, ok := o.(StackPolicyAttacher)
+	if !ok {
+		return nil, nil
```
How about I move the check of satisfying the interface to before the func call? ... I feel this would illustrate the intension (that only things that assign policies need to request a role name) more clearly?

Commit:
clarify optional PolicyAttacher/SecretOutputter

implmenting StackPolicyAttacher (to request a role name to attach a
policy to) or StackSecretNamer (to name a Secret that should be
populated with cloudformation outputs) is optional. The code was not
making this clear previously as the casts were being attetmpted at a
point that made them look like unchecked error conditions. By checking
the codition before the call the intension is hopefully clarified.

Commit:
check principal in READY state before policy

there is a race between provisioning of "Prinicals" (roles) and
StackPolicyAssigners (policy)

to address this race, we wait for a READY state on the principal record
before attempting to provision things that want a role name.

Commit:
removes unused constants

some consts were left laying around... remove them and make naming of
others more consistant where sensible

Commit:
ignore generated manifest files

the generated manifest files (crds/webhooks) are currently an
intermediate step before generating the chart and we do not commit them
to the repo.

Commit:
update service operator crds

some new fields have been added and the status field changed for the
service operator types so the CRDs have been regenerated with `make
update-chart`

blairboy362 reviewed (dismissed):
I'm sure it's fine

Commit:
add GetRolePolicy for service-operator role

GetRolePolicy is required by cloudformation to avoid assigning clashing
policies.

Commit:
make all status fields optional

we want to do "best effort" collecting of any status information on
resource status as this allows us to collect as much data as possible
when there is an issue.

this means we cannot relie on the status fields always having non-zero
values, and they should be optional in the CRD

Commit:
only update events if we have events

after a stack is in a rollback_complete or deleted state then the events
disappear... we want to keep the events around as it's useful for
debugging, so only wipe out / reset the status events if we actually
have some to update, otherwise leave the old ones

Commit:
add instanceCount:1 to postgres example

Commit:
alwways update status even if reconcile errors

after running reconcile we attempt to update any changes we have made to
our resource object.

if the reconcile errors, we _still_ want to try the update, but
returning an err from the CreateOrUpdate mutate func will cause it to
bail out, so instead we should check the errors seperately to ensure
update is attempted.

Commit:
add missing error state mapping

Commit:
lower reconcile timeout to 1min

the reconcile timeout dictates how long the reconcile func will wait for
cloudformation stack to get into the state it wants. Too low and it will
require the api-server to requeue to reconcile too often and will get
automatically rate limited too long and we won't see status updates on
the objects frequently enough.... between 1 and 5 mins are probably
sensible values.

Commit:
typo

Co-Authored-By: Alex Monk <alex.monk@digital.cabinet-office.gov.uk>

Krenair commented on charts/gsp-cluster/templates/02-gsp-system/service-operator/crd/access.govsvc.uk_principals.yaml:
```diff
@@ -28,46 +28,51 @@ spec:
+              properties:
+                events:
+                  description: Events will hold more in-depth details of the current
+                    state of the instance.
```
sure

Krenair commented on modules/gsp-cluster/service-operator.tf:
```diff
@@ -35,6 +35,16 @@ data "aws_iam_policy_document" "service-operator" {
+    ]
+
+    resources = [
+      "*",
```
Do we not get even a known prefix of ARNs?

chrisfarms replied:
not that is under our control, no

Krenair commented:
Please be careful with force-pushes on this PR. Thankfully GitHub's 'force-pushed' link shows a diff of the changes made and it looks like just a rebase without any additional relevant changes.

Krenair commented on components/service-operator/apis/database/v1beta1/postgres_types.go:
```diff
@@ -16,60 +16,213 @@ limitations under the License.
+	}
+
+	// create a reference to the values created in secrets manager
+	masterUsernameSecretRef := cloudformation.Join("", []string{"{{resolve:secretsmanager:", cloudformation.Ref(PostgresResourceMasterCredentials), ":SecretString:username}}"})
```
Expecting this SecretsManager resolve thing to work in a CloudFormation Join isn't something I'd necessarily expect to work, but if it does work, ok.

Krenair commented on components/service-operator/apis/database/v1beta1/postgres_types.go:
```diff
@@ -16,60 +16,213 @@ limitations under the License.
+		"Value":       masterUsernameSecretRef,
+	}
+
+	template.Outputs[PostgresPassword] = map[string]interface{}{
```
This is all sorts of scary. Can we trust that secrets manager will never try to interpret the resolve syntax in here?

chrisfarms replied:
yes, it's a magic string thanks to secretsmanager

Krenair commented on components/service-operator/controllers/principal_cloudformation_test.go:
```diff
@@ -0,0 +1,129 @@
+			}).Should(ContainSubstring("xxx-principal-test-test-role"))
+		})
+
+		By("ensuring a finalizaer is present on resource to prevent deletion", func() {
```
finalizaer is a typo across all three controller tests

Krenair commented on components/service-operator/controllers/principal_cloudformation_test.go:
```diff
@@ -0,0 +1,129 @@
+			}).Should(BeTrue())
+		})
+
+		By("deleteing resource with kubernetes api", func() {
```
deleteing is a typo across all three controller tests

Krenair commented on components/service-operator/controllers/sqs_cloudformation_test.go:
```diff
@@ -0,0 +1,170 @@
+			}).Should(HaveKey("QueueURL"))
+		})
+
+		By("creating a secret with the prinical role name", func() {
```
principal

Krenair commented on components/service-operator/controllers/postgres_cloudformation_test.go:
```diff
@@ -0,0 +1,163 @@
+			Eventually(func() map[string][]byte {
+				_ = client.Get(ctx, secretNamespacedName, &secret)
+				return secret.Data
+			}).Should(HaveKey("Endpoint"))
```
Also we probably want one for ReadEndpoint

chrisfarms replied:
Gonna add these into unit tests for GetStackTemplate on the types in future PR

Krenair commented on components/service-operator/controllers/sqs_cloudformation_test.go:
```diff
@@ -0,0 +1,170 @@
+			}, timeout).Should(Equal(0))
+		})
+
+		// By("ensuring secret has been removed", func() {
```
comment from the others: `// GC will remove this in a real cluster, but we don't have the hooks installed in our tests :(`

Krenair commented on components/service-operator/internal/object/status.go:
```diff
@@ -0,0 +1,91 @@
+var _ StatusWriter = &Status{}
+
+// GetStatus returns the status
+func (s *Status) GetStatus() Status {
```
This seems interesting - it returns... itself? Or a new object if it itself is nil?

chrisfarms replied:
it doesn't return a pointer to itself ... it returns a copy of it's status or an empty status .... this is simply implementing the object.StatusReader interface

Krenair commented on components/service-operator/controllers/types.go:
```diff
@@ -0,0 +1,65 @@
+	return r.Reconciler.SetupWithManager(mgr)
+}
+
+// Err shifts the first error off the list of collected errors, returns it and flushes the error queue
```
It just seems to get the first element if there is one, otherwise returns nil? `Errs` below flushes the queue.

chrisfarms replied:
correct.

Krenair commented on components/service-operator/internal/aws/cloudformation/types.go:
```diff
@@ -0,0 +1,41 @@
+	GetStackOutputWhitelist() []string
+}
+
+// StackSecretOutputter allows a type to return the name of a kubernetes Secret
```
I thought we said outputs can never be sensitive.

Krenair replied:
For the record we discussed this at length and basically it comes down to the distinction between 'sensitive' and 'secret'. In this particular case the definition of sensitive includes usernames, account numbers, URLs. We choose not to return actual *secrets* such as passwords in our stacks - those get returned as references which we then have to resolve using secrets manager elsewhere in the service operator, and this is still useful for that.

Krenair commented on components/service-operator/internal/aws/cloudformation/client.go:
```diff
@@ -0,0 +1,477 @@
+			if IsNotFoundError(err) && in(DeleteComplete, desiredStates) {
+				// If we are waiting for DeleteComplete state and the
+				// stack has gone missing, consider this DeleteComplete
+				return &State{}, nil
```
Shouldn't this be a string like DELETE_COMPLETE, or DeleteComplete?

chrisfarms replied:
This function's job is to wait for a state and then return the entire cloudformation stack data object (State) ... When the stack has been deleted we don't have a stack data object, but it's also not an error... The best it can do is return a blank data object which is slightly more representative than nil

It is a bit weird ... But the DELETE_COMPLETE string will only show for about a second before vanishing so can't be trusted.

The cloudformation.State type/alias is a bit of a confusing name, but it was slightly less confusing than two different "Stack" types

Krenair replied:
Ohhh... It's actually an empty stack. Let's see if we can find a better alternative to `type State = cloudformation.Stack` on Monday.

Krenair replied:
We'll return to this in a later PR.

Krenair commented on components/service-operator/internal/aws/cloudformation/client.go:
```diff
@@ -0,0 +1,477 @@
+		} else if awsErr.Code() == "ValidationError" && strings.Contains(awsErr.Message(), NoExistErrMatch) {
+			return true
+		}
+	}
```
I kind of wonder if we should say that if it's not an awserror.Error then return false?

Krenair replied:
What I said makes literally no sense and I no longer follow myself.

chrisfarms commented on components/service-operator/internal/aws/cloudformation/client.go:
```diff
@@ -0,0 +1,477 @@
+			}
+			// phew found it
+			if *param.ParameterKey == wantedKey {
+				delete(missing, wantedKey)
```
sure why not, save those nanoseconds :)

Commit:
break an save a cycle or two

Commit:
adds sanity check for DescribeStacks returning >1

if for some weird reason DescribeStacks returned more than 1 result,
then something has gone wrong with the AWS SDK client (or more likely,
something has gone wrong with the fake mock implementation of the
client).

chrisfarms commented on components/service-operator/internal/aws/cloudformation/client.go:
```diff
@@ -0,0 +1,477 @@
+	if describeOutput == nil {
+		return nil, fmt.Errorf("describeOutput was nil, potential issue with AWS Client")
+	}
+	if len(describeOutput.Stacks) == 0 {
```
these three errors are catching things that the AWS SDK documentation say shouldn't happen (and mainly exist to avoid a `panics` that arrose when creating the fake aws mock). checking for >1 seems like a sensible sanity check, I'll add that.

chrisfarms replied:
done
